content/blog/2023/11/2023-11-29-jenkins-to-github.adoc

```diff
@@ -0,0 +1,270 @@
+Then, you https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app-installation[call a GitHub endpoint] with this JWT, saying "I'm this App, and I'm trying to access the repos we agreed upon earlier." This endpoint gives you an "installation access token" (IAT).
+Your script would use this IAT to call other endpoints, just like how you'd use a PAT when calling endpoints manually as yourself.
+
+Luckily, many languages have libraries that handles the intermediate work for you.
+Let's say your Jenkins file runs a Python script via (`sh './run.py'`).
+We can use the https://github.com/PyGithub/PyGithub[PyGitHub] library to call GitHub APIs Pythonically.
+
+In Python, the built-in https://docs.python.org/3/library/logging.html[logging] library is the standard way to emit messages.
+Your script would create a "logger" and attaches different "handlers" to it, each defining an output channel of your messages -- standard output, file on disk, remote service, etc.
+Intuitively, we can write a handler for logging to GitHub Checks.
+
+With the help of ChatGPT, I implemented one https://gist.github.com/tslmy/84f34a25babe045eb302ec72f2bf39eb[here].
+You can see that the handler requires quite some parameters to initialize.
+Where do you get them?
+
+* For locating the PR and the commit, the https://plugins.jenkins.io/github-pullrequest/[GitHub Integration Plugin] can help.
+It can trigger Jenkins jobs upon PR events, injecting https://github.com/KostyaSha/github-integration-plugin/blob/master/docs/Configuration.adoc#available-environment-variables[these environment variables] to each Build.
+We'll use two: `GITHUB_PR_HEAD_SHA` and `GITHUB_REPO_SSH_URL`.
+* For referencing from the PR Check to the Jenkins Build, we'll also need an env.
+var.
+set by vanilla Jenkins: `BUILD_URL`.
+You can find a list of all Jenkins-set variables https://wiki.jenkins.io/display/JENKINS/Building+a+software+project#Buildingasoftwareproject-belowJenkinsSetEnvironmentVariables[here].
+* For authentication, we are on our own.
+We can upload the App's private key to Jenkins credentials storage and wrap the `sh './run.py'` step with the closure `sshUserPrivateKey`.
+For the App ID and the Installation ID, since they aren't necessarily secrets, we can afford to spell them out with an `environment` directive.
+Combined, your Jenkinsfile might look like this:
++
+[,groovy]
+----
+pipeline {
+environment {
+  GITHUB_APP_ID = '123'
+  GITHUB_APP_INSTALLATION_ID = '456'
+}
+stages {
+  stage('...') {
+    steps {
+      script {
+        sshUserPrivateKey (
+          credentialsId: "...", // ID from the Jenkins credentials storage.
+          keyFileVariable: 'GITHUB_APP_PRIVATE_KEY_PATH') {
+            sh './run.py'
+----
+
+Now, in your Python script, you can read the env.
+var.s during initialization:
```

Comment:
```suggestion
Now, in your Python script, you can read the environment variables during initialization:
```
Suggestion to use environment variables so that the full stop is not configured incorrectly. Alternatively, the proper shorthand can be used instead.